translates: tutorial (broken replaced by #73) (#29)
reactjs/nl.reactjs.org: content/tutorial/tutorial.md

  - "docs/tutorial-zh-CN.html"
---

> Deze tutorial vereist geen voorbestaande React kennis.

Reviewer: Should we translate the word `tutorial` as well? An option for this would be `inleiding` (introduction).

```suggestion
Deze inleiding vereist geen voorbestaande React kennis.
```

Do we have any guidelines for this? Can we/should we start a glossary?

Reviewer 2: We have a (partial) glossary #5

Reviewer 3: > Should we translate the word `tutorial` as well? An option for this would be `inleiding` (introduction).

`tutorial` is a [valid Dutch (loan)word](https://www.vandale.nl/gratis-woordenboek/nederlands/betekenis/tutorial). However if we insist on translating it, maybe `les` (lesson) would be a better translation than `inleiding`?

Reviewer 4: Since `tutorial` isn't yet included in the glossary #5 , I'd be in favor of adding it to the keep as-is list. I think part of the meaning is lost in translation with `inleiding`, `les`, etc.

Reviewer 5: assume => gaat niet uit van
voorbestaande kennis => voorkennis

```suggestion
Deze tutorial gaat niet uit van enige voorkennis van React.
```

> ## Voordat we beginnen {#voordat-we-beginnen}

Reviewer: I don't think we should translate the anchors, this way they'll still work when switching languages. I looked up an example where this also wasn't done for the Spanish translation: https://github.com/reactjs/es.reactjs.org/pull/199/files#diff-c9112233168e6093109b86f0f5b0f328R16

```suggestion
## Voordat we beginnen {#before-we-start-the-tutorial}
```

Reviewer 2: We shouldn't no

> ## Voordat we beginnen {#before-we-start-the-tutorial}

Reviewer 4: ```suggestion
## Voordat We Beginnen {#before-we-start-the-tutorial}
```

> We zullen een klein spelletje bouwen in deze tutorial. **Misschien ben je geneigd om dit over te slaan aangezien je geen games maakt -- maar probeer het.** De technieken die je zal leren in deze tutorial zijn fundamenteel om React applicaties te bouwen, eens je deze onder de knie krijgt, heb je een diep inzicht in React. 

Reviewer: Replacing `tutorial` with `inleiding` for consistency. Also suggesting `maar geef het een kans` to stick closer to the original intention and have a better flowing sentence.

I do like the usage of `eens` but we could possibly simplify to `wanneer` (omitted in the suggestion for your discretion).

```suggestion
We zullen een klein spelletje bouwen in deze inleiding. **Misschien ben je geneigd om dit over te slaan aangezien je geen games maakt -- maar geef het een kans.** De technieken die je zal leren in deze inleiding zijn fundamenteel om React applicaties te bouwen, eens je deze onder de knie krijgt, heb je een diep inzicht in React. 
```

>Tip
>

> >Deze tutorial is bedoelt voor degene die het liefst **al doende leren**. Indien je liever concepten vanuit de basis leert, neem dan een kijkje naar de [stapsgewijze documentatie](/docs/hello-world.html).

Reviewer: `degene` is the singular form so then it should be `al doende leert`. I do think keeping it plural here is fine so then we'll move to `degenen` (https://onzetaal.nl/taaladvies/degene-diegene). Also moving `tutorial` to `inleiding`

```suggestion
>Deze inleiding is bedoelt voor degenen die het liefst **al doende leren**. Indien je liever concepten vanuit de basis leert, neem dan een kijkje naar de [stapsgewijze documentatie](/docs/hello-world.html).
```

> De tutorial is opgedeeld in verschillende onderdelen:

Reviewer: `De` to `Deze` to remain consistent with other parts of the text.

```suggestion
Deze inleiding is opgedeeld in verschillende onderdelen:
```

> * [Setup voor de Tutorial](#setup-for-the-tutorial) geeft je *een beginpunt* om de tutorial te volgen.

Reviewer: ```suggestion
* [Setup voor de Inleiding](#setup-for-the-tutorial) geeft je *een beginpunt* om de inleiding te volgen.
```

> * [Setup voor de Tutorial](#setup-for-the-tutorial) geeft je *een beginpunt* om de tutorial te volgen.

Reviewer 4: ```suggestion
* [Setup Voor De Tutorial](#setup-for-the-tutorial) geeft je **een beginpunt** om de tutorial te volgen.
```

* [Overzicht](#overview) leert je de **fundamentele concepten** van React: componenten (components), eigenschappen (props), en staat (state).

> * [Het spel afwerken](#completing-the-game) leert je de **meest voorkomende technieken** in React development.

Reviewer 4: ```suggestion
* [Het Spel Afwerken](#completing-the-game) leert je de **meest voorkomende technieken** in React ontwikkeling.
```

* [Tijdreizen toevoegen](#adding-time-travel) zal je **een dieper inzicht** geven in de unieke voordelen van React

> Het is niet noodzakelijk om alle onderdelen in één keer te vervolledigen. Probeer zo ver mogelijk te geraken -- zelfs al is het een of twee onderdelen.

Reviewer: Simplify

```suggestion
Het is niet noodzakelijk om alle onderdelen in één keer te voltooien. Probeer zo ver mogelijk te komen -- zelfs al is het een of twee onderdelen.
```

> Je kan code copy-pasten terwijl je de tutorial volgt, maar we raden je aan om de code zelf uit te typen. Dit helpt je om gewoontes te creëren en geeft je een dieper inzicht. 

Reviewer: ```suggestion
Je kan code kopiëren en plakken terwijl je de inleiding volgt, maar we raden je aan om de code zelf uit te typen. Dit helpt je om gewoontes te creëren en geeft je een dieper inzicht. 
```

> ### Wat gaan we maken? {#wat-gaan-we-maken}

Reviewer: Keeping anchors in English for cross translation linking
```suggestion
### Wat gaan we maken? {#what-are-we-building}
```

> ### Wat gaan we maken? {#what-are-we-building}

Reviewer 4: ```suggestion
### Wat Gaan We Maken? {#what-are-we-building}
```

> In deze tutorial, zullen we tonen hoe een interactieve tic-tac-toe (boter-kaas-eieren) spel maken met React.

Reviewer: ```suggestion
In deze inleiding, zullen we tonen hoe je een interactief boter-kaas-en-eieren spel maakt met React.
```

> In deze tutorial, zullen we tonen hoe een interactieve tic-tac-toe (boter-kaas-eieren) spel maken met React.

Reviewer 4: ```suggestion
In deze tutorial, zullen we tonen hoe je een interactief tic-tac-toe (boter-kaas-eieren) spel kan maken met React.
```

> Je kan een kijkje nemen naar wat we zullen maken: **[Afgewerkt Resultaat](https://codepen.io/gaearon/pen/gWWZgR?editors=0010)**. Als de code nog geen steek houdt of je syntax nog niet kent, is dat geen probleem. Het doel van deze tutorial is je React en de syntax aan te leren.

Reviewer: ```suggestion
Je kan een kijkje nemen naar wat we zullen maken: **[Afgewerkt Resultaat](https://codepen.io/gaearon/pen/gWWZgR?editors=0010)**. Als de code nog geen steek houdt of je syntax nog niet kent, is dat geen probleem. Het doel van deze inleiding is je React en de syntax aan te leren.
```

> We raden aan dat een kijkje neemt naar het tic-tac-toe spel voordat je verdergaat met de tutorial. Een van de eigenschappen dat je zal opmerken is dat er een genummerde lijst is aan de rechterkant van het spel. Deze lijst geeft je een geschiedenis van alle zetten die in het spel zijn voorgekomen, en deze wordt geupdate tijdens het spel.

Reviewer: ```suggestion
We raden aan dat je een kijkje neemt naar het boter-kaas-en-eieren spel voordat je verdergaat met de inleiding. Een van de eigenschappen die je zal opmerken is dat er een genummerde lijst is aan de rechterkant van het spel. Deze lijst geeft je een geschiedenis van alle zetten die in het spel zijn voorgekomen, en deze wordt geüpdatet tijdens het spel.
```

> We raden aan dat een kijkje neemt naar het tic-tac-toe spel voordat je verdergaat met de tutorial. Een van de eigenschappen dat je zal opmerken is dat er een genummerde lijst is aan de rechterkant van het spel. Deze lijst geeft je een geschiedenis van alle zetten die in het spel zijn voorgekomen, en deze wordt geupdate tijdens het spel.

Reviewer 4: ```suggestion
We raden aan dat je het tic-tac-toe spel eens bekijkt voordat je verdergaat met de tutorial. Eén van de eigenschappen die je zal opmerken is dat er een genummerde lijst is aan de rechterkant van het spel. Deze lijst geeft je een geschiedenis van alle zetten die in het spel zijn voorgekomen, en deze wordt geüpdatet tijdens het spel.
```

Eens je vertrouwd bent met het tic-tac-toe spel, kan je het afsluiten. We zullen vanuit een simpelere template starten voor deze tutorial. 
Onze volgende stap is jouw setup in te stellen zodat je het spel kan beginnen te bouwen.

### Vereisten {#prerequisites}

> We gaan ervanuit dat je enigzins vertrouwd bent met HTML en Javascript, maar het is mogelijk om deze tutorial te volgen, zelfs als je van een andere programmeer taal komt. We gaan er ook vanuit dat je vertrouwd bent met programmeer concepten zoals functies (functions), objecten (objects), arrays en in mindere mate met klasses (classes).

Reviewer 4: ```suggestion
We gaan ervan uit dat je enigszins vertrouwd bent met HTML en JavaScript, maar het is mogelijk om deze tutorial te volgen, zelfs als je van een andere programmeertaal komt. We gaan er ook van uit dat je vertrouwd bent met programmeerconcepten zoals functies (functions), objecten (objects), arrays en in mindere mate met klasses (classes).
```

> Als je Javascript wil herhalen, raden we [deze handleiding](https://developer.mozilla.org/en-US/docs/Web/JavaScript/A_re-introduction_to_JavaScript).

Reviewer 4: ```suggestion
Als je JavaScript wil herhalen, raden we [deze handleiding](https://developer.mozilla.org/en-US/docs/Web/JavaScript/A_re-introduction_to_JavaScript) aan.
```

We gebruiken ook sommige functies van ES6 -- een recente versie van Javscript. In deze tutorial zullen we [arrow functies](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions), [classes](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes), [`let`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/let), en [`const`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/const) statements gebruiken.

> Je kan de [Babel REPL](babel://es5-syntax-example) gebruiken om de controleren hoe ES6 code compileert.

Reviewer 4: ```suggestion
Je kan de [Babel REPL](babel://es5-syntax-example) gebruiken om te controleren hoe ES6 code compileert.
```

> ## Setup voor de Tutorial {#setup-for-the-tutorial}

Reviewer 4: ```suggestion
## Setup Voor De Tutorial {#setup-for-the-tutorial}
```

There are two ways to complete this tutorial: you can either write the code in your browser, or you can set up a local development environment on your computer.

> ### Setup Optie 1: Schrijf Code in de Browser {#setup-option-1-write-code-in-the-browser}

Reviewer 4: ```suggestion
### Setup Optie 1: Schrijf Code In De Browser {#setup-option-1-write-code-in-the-browser}
```

Dit is de snelste manier om te starten!

> Eerst, open deze **[Starter Code](https://codepen.io/gaearon/pen/oWWQNa?editors=0010)** in een nieuwe tab. De nieuwe tab toont een leeg tic-tac-toe bord en React Code. We zullen de React code aanpassen in deze tutorial.

Reviewer 4: ```suggestion
Open eerst deze **[Start Code](https://codepen.io/gaearon/pen/oWWQNa?editors=0010)** in een nieuwe tab. De nieuwe tab zou een leeg tic-tac-toe bord en React Code moeten weergeven. We zullen de React code aanpassen in deze tutorial.
```

> Je kan de tweede setup optie overslaan en dadelijk naar het [overzicht](#overview) hoofstuk voor een overzicht van React.

Reviewer 4: ```suggestion
Je kan nu de tweede setup optie overslaan en dadelijk naar het [overzicht](#overview) hoofdstuk gaan voor een overzicht van React.
```

> ### Setup Optie 2: Local Development Environment {#setup-option-2-local-development-environment}

Reviewer 4: ```suggestion
### Setup Optie 2: Lokale Ontwikkelingsomgeving {#setup-option-2-local-development-environment}
```

Dit is volledig optioneel en niet vereist voor deze tutorial!
<br>

<details>

<summary><b>Optioneel: Instructies om lokaal te volgen met je favoriete text editor</b></summary>

Deze setup vereist meer werk maar laat je toe om de volledige tutorial af te werken met je favoriete text editor. Hier zijn de stappen om te volgen:

1. Zorg ervoor dat je een recente versie van [Node.js](https://nodejs.org/en/) geïnstalleerd hebt.
2. Volg de [installatie instructies voor Create React App](/docs/create-a-new-react-app.html#create-react-app) om een nieuw project aan te maken.

```bash
npx create-react-app my-app
```

3. Verwijder alle bestanden in de map `src/` van het nieuwe project.

> > Nota:

Reviewer 4: ```suggestion
> Opmerking:
```

>
>**Don't delete the entire `src` folder, just the original source files inside it.** We'll replace the default source files with examples for this project in the next step.

> **Verwijder niet de gehele `src` map, enkel de bestanden in de map (de source files)** We zullen voor dit project de standaard bestanden met voorbeelden vervangen in de volgende stap.

Reviewer 4: ```suggestion
>**Verwijder niet de gehele `src` map, maar enkel de originele source files in de map.** We zullen de standaard bestanden met voorbeelden voor dit project vervangen in de volgende stap.
```

```bash
cd my-app
cd src

# Voor Mac of Linux gebruikers:
rm -f *

# Of voor Windows gebruikers:
del *

# Ga dan terug naar de project map
cd ..
```

4. Voeg een bestand toe met de naam `index.css` in de map `src/` met [deze CSS code](https://codepen.io/gaearon/pen/oWWQNa?editors=0100).

5.  Voeg een bestand toe met de naam `index.js` in de map `src/` met  [deze JS code](https://codepen.io/gaearon/pen/oWWQNa?editors=0010).

6. Voeg deze drie lijnen toe aan het begin van `index.js` in de map `src`:

```js
import React from 'react';
import ReactDOM from 'react-dom';
import './index.css';
```

> Als je nu `npm start` in de map van het project gebruikt en `http://localhost:3000` in je browser opent, zie je een leeg tic-tac-toe veld.

Reviewer 4: ```suggestion
Als je nu `npm start` in de map van het project uitvoert en `http://localhost:3000` in je browser opent, zou je een leeg tic-tac-toe veld moeten zien.
```

> We raden [deze instructies](https://babeljs.io/docs/editors/) om de *syntax highlighting* in te stellen in je editor.

Reviewer 4: ```suggestion
We raden [deze instructies](https://babeljs.io/docs/editors/) om de syntax highlighting in te stellen in je editor.
```

</details>

> ### Help, ik zit vast! {#help-im-stuck}

Reviewer 4: ```suggestion
### Help, Ik Zit Vast! {#help-im-stuck}
```

If you get stuck, check out the [community support resources](/community/support.html). In particular, [Reactiflux Chat](https://discord.gg/0ZcbPKXt5bZjGY5n) is a great way to get help quickly. If you don't receive an answer, or if you remain stuck, please file an issue, and we'll help you out.

> Als je vast zit, ga dan naar de [community support resources](/community/support.html).  [Reactiflux Chat](https://discord.gg/0ZcbPKXt5bZjGY5n) in het bijzonder is een geweldige manier om snel hulp te verkrijgen. Als je geen antwoord krijgt, of je vast blijft te zitten, open dan een *issue*, en we helpen je verder.

Reviewer 4: ```suggestion
Als je vast zit, ga dan naar de [community support resources](/community/support.html). [Reactiflux Chat](https://discord.gg/0ZcbPKXt5bZjGY5n) in het bijzonder is een geweldige manier om snel hulp te verkrijgen. Als je geen antwoord krijgt, of als je vast blijft zitten, open dan een *issue*, en we helpen je verder.
```

## Overzicht {#overview}

> Nu dat je setup klaar is, tijd voor een overzicht van React!

Reviewer 4: ```suggestion
Nu je setup klaar is, is het tijd voor een overzicht van React!
```

### Wat Is React? {#what-is-react}

> React is een declaratieve, efficiente, en flexibele Javascript bibliotheek (library) om gebruikersinterfaces (user interfaces of UI) te bouwen. Het laat je toe om complexe UIs te maken met kleine en geisoleerde deeltjes code die componenten (components) genoemd worden.

Reviewer 4: ```suggestion
React is een declaratieve, efficiënte, en flexibele JavaScript bibliotheek (library) om gebruikersinterfaces (user interfaces of UI) te bouwen. Het laat je toe om complexe UIs te maken met kleine en geïsoleerde deeltjes code die componenten (components) genoemd worden.
```

> React heeft verschillende soorten componenten, maar we zullen starten met de `React.Component` subklasse:

Reviewer 4: ```suggestion
React heeft verschillende soorten componenten, maar we zullen starten met subklasses van `React.Component`:
```

// Example usage: <ShoppingList name="Mark" />
```

> We komen zodadelijk tot de rare XML-achtige tags. We gebruiken componenten om React te vertellen wat we op het scherm willen zien. Wanneer onze data verandert, zal React op een efficiente manier onze componenten updaten en opnieuw renderen.

Reviewer 4: ```suggestion
We komen zo dadelijk tot de rare XML-achtige tags. We gebruiken componenten om React te vertellen wat we op het scherm willen zien. Wanneer onze data verandert, zal React op een efficiënte manier onze componenten updaten en opnieuw renderen.
```

> In het voorbeeld is ShoppingList een **React component class**, of ook ** React component type**. Een component krijgt parameter, `props` (kort voor "properties") genaamd en returned een hierarchie van views om weer te geven via de `render` methode.

Reviewer 4: ```suggestion
In het voorbeeld is ShoppingList een **React component class**, of ook **React component type**. Een component krijgt parameters, `props` (kort voor "properties") genaamd en geeft een hiërarchie van views terug om weer te geven via de `render` methode.
```

> De `render` methode gebruikt een *beschrijving* van wat je wilt zien om het scherm. React neemt deze beschrijving en geeft het resultaat weer. In het bijzonder, `render` returned een **React element**, een lichte beschrijving van wat te renderen. De meeste React ontwikkelaars gebruiken een speciale syntax "JSX" dat het toelaat om deze structuren gemakkelijker schrijven. De `<div />` syntax wordt tijdens de *build time* omgezet naar `React.createElement('div')`. 

Reviewer 4: ```suggestion
De `render` methode gebruikt een *beschrijving* van wat je wilt zien op het scherm. React neemt deze beschrijving en geeft het resultaat weer. In het bijzonder, `render` geeft een **React element** terug, een lichte beschrijving van wat te renderen. De meeste React ontwikkelaars gebruiken een speciale syntax genaamd "JSX" die het toelaat om deze structuren gemakkelijker schrijven. De `<div />` syntax wordt tijdens *build time* omgezet naar `React.createElement('div')`. 
```

);
```

[De volledige uitgebreide versie.](babel://tutorial-expanded-version)

Als je nieuwsgierig bent, `createElement()` wordt in meer detail omschreven in de [API referentie](/docs/react-api.html#createelement), maar we zullen het niet gebruiken in deze tutorial. In de plaats hiervan zullen we JSX blijven gebruiken.

> JSX komt met de volledige functionaliteit van Javascript. Je kan *elke* Javascript expressie gebruiken binnen haakjes in JSX. Elk React element is een Javascript object dat je kan opslaan in een variabele of doorgeven in je programma.

Reviewer 4: ```suggestion
JSX komt met de volledige functionaliteit van JavaScript. Je kan *elke* JavaScript expressie gebruiken binnen haakjes in JSX. Elk React element is een JavaScript object dat je kan opslaan in een variabele of doorgeven in je programma.
```

> Het `ShoppingList` component van hierboven rendert enkel in de ingebouwde DOM componenten zoals `<div />` en `<li />`. Maar je kan ook custom React componenten aanmaken en renderen. Bijvoorbeeld. we kunnen nu naar de hele *shopping list* refereren met `<ShoppingList />`.

Reviewer 4: ```suggestion
Het `ShoppingList` component van hierboven rendert enkel de ingebouwde DOM componenten zoals `<div />` en `<li />`. Maar je kan ook custom React componenten aanmaken en renderen. Bijvoorbeeld, we kunnen nu naar de hele *shopping list* verwijzen met `<ShoppingList />`.
```

Elk React component is ingekapseld en kan onafhankelijk bestaan; dit laat ons toe om complexe UIs te maken van eenvoudige componenten.

## De Start Code Inspecteren {#inspecting-the-starter-code}

> Als je **in je browser** aan de tutorial zal werken, open deze code in een nieuwe tab: **[Start Code](https://codepen.io/gaearon/pen/oWWQNa?editors=0010)**. Als je **lokaal** aan de tutorial werkt, open dan `src/index.js` in je project map (je hebt dit bestand reeds aangemaakt tijdens de [setup](#setup-option-2-local-development-environment)).

Reviewer 4: ```suggestion
Als je **in je browser** aan de tutorial gaat werken, open dan deze code in een nieuwe tab: **[Start Code](https://codepen.io/gaearon/pen/oWWQNa?editors=0010)**. Als je **lokaal** aan de tutorial gaat werken, open dan `src/index.js` in je project map (je hebt dit bestand reeds aangemaakt tijdens de [setup](#setup-option-2-local-development-environment)).
```

Reviewer 3: updated for clarity

> Deze Start Code is de basis van wat we gaan maken. Wij hebben reeds de CSS styling bezorgd zodat je je enkel hoeft te concentreren op het leren van React en het programmeren van het tic-tac-toe spel

Reviewer 4: ```suggestion
Deze Start Code is de basis van wat we gaan maken. We hebben reeds voor de CSS styling gezorgd zodat je je enkel hoeft te concentreren op het leren van React en het programmeren van het tic-tac-toe spel.
```

Bij het inspecteren van de code zal je opmerken dat we drie React componenten hebben:

* Square
* Board
* Game

> Het Square component rendert een `<button>` en het Board component rendert 9 vierkante vakken Het Game component rendert een bord met een and the Board renders 9 squares met placeholder waardes die we later zullen aanpassen. Momenteel zijn er geen interactieve componenten.

Reviewer 4: ```suggestion
Het Square component rendert een `<button>` en het Board component rendert 9 vierkante vakken. Het Game component rendert een bord met plaatshoudende waarden die we later zullen aanpassen. Momenteel zijn er geen interactieve componenten.
```

> ### Data doorgeven door middel van Props {#passing-data-through-props}

Reviewer 4: ```suggestion
### Data Doorgeven Met Props {#passing-data-through-props}
```

> Om te beginnen, proberen we data van het Board component naar het Square component door te geven.

Reviewer 4: ```suggestion
Om te beginnen, zullen we proberen data van het Board component aan het Square component door te geven.
```

  }
```

>Nota
>
> We splitsen het gereturnde element in verschillende lijnen voor leesbaarheid, en we voegen haakjes toe zodat Javascript geen puntcomma invoegt na `return` en zo onze code breekt.

> Nu zijn we twee props door aan het geven van Board naar Square: `value` en `onClick`. De `onClick` prop is een functie die het Square component can oproepen wanneer het aangeklikt wordt. We maken de volgende veranderingen aan Square:

Reviewer 4: ```suggestion
Nu zijn we twee props door aan het geven van Board naar Square: `value` en `onClick`. De `onClick` prop is een functie die het Square component kan aanroepen wanneer het aangeklikt wordt. We maken de volgende veranderingen aan Square:
```

* Vervang `this.state.value` met `this.props.value` in de Square `render` methode
* Vervang `this.setState()` met `this.props.onClick()` in de Square `render` methode

> * Delete de `constructor` van Square want Square houdt niet langer de staat van het spel bij

Reviewer 4: ```suggestion
* Verwijder de `constructor` van Square want Square houdt niet langer de staat van het spel bij
```

}
```

> Wanneer een Square aangeklikt wordt, wordt de functie `onClick`, die door Board bezorgd is opgeroepen. Hier is een overzicht van hoe dit wordt toegepast:

Reviewer 4: ```suggestion
Wanneer een Square aangeklikt wordt, wordt de functie `onClick`, die door Board bezorgd is, aangeroepen. Hier is een overzicht van hoe dit wordt bereikt:
```

1. De `onClick` prop in het ingebouwde DOM `<button>` component zegt aan React om een click event listener op te zetten.
2. Wanneer de knop aangeklikt wordt, roept React de `onClick` event handler die gedefinieerd is in de `render()` methode van Square.
3. Deze event handler roept `this.props.onClick()` op. De prop `onClick` prop van Square is gespecifieerd door Board.
4. Aangezien Board `onClick={() => this.handleClick(i)}` aan Square doorgaf, roept Square `this.handleClick(i)` wanneer het aangeklikt wordt.
5. We hebben de methode `handleClick()` nog niet gedefinieerd dus onze code crashed.

> >Nota

Reviewer 4: ```suggestion
>Opmerking
```

}
```

**[De volledige code tot zover](https://codepen.io/gaearon/pen/ybbQJX?editors=0010)**

> Na deze veranderingen zijn hebben we de mogelijkheid weer om de Square componenten aan te klikken om deze op te vullen. Maar nu is de state in het Board component opgeslagen in plaats van in de individuele Square componenten. Wanneer de state van Board verandert zullen de Square componenten automatisch her-renderen. Het opslaan van de state van alle Square componenten in het Board component staat toe om later een winner uit te roepen.

Reviewer 4: ```suggestion
Na deze veranderingen kunnen we de Square componenten weer aanklikken om deze op te vullen. Maar nu is de state in het Board component opgeslagen in plaats van in de individuele Square componenten. Wanneer de state van Board verandert zullen de Square componenten automatisch opnieuw renderen. Het opslaan van de state van alle Square componenten in het Board component staat toe om later een winnaar te determineren.
```

> Aangezien de Square componenten niet langer de state bevatten, de Square componenten waardes ontvangen van het Board component en het Board component informeren wanneer ze aangeklikt worden. In de React terminologie, zijn deze nu **controlled components**. Het Board component heeft volledige controle over hen.

Reviewer 4: ```suggestion
Aangezien de Square componenten niet langer de state bevatten, de Square componenten waardes ontvangen van het Board component en het Board component informeren wanneer ze aangeklikt worden. Volgens React terminologie, zijn deze nu **controlled components**. Het Board component heeft volledige controle over hen.
```

> Merk op hoe in `handleClick` we `.slice()` oproepen om een kopie van de `squares` array te modifieeren in plaats van de bestaande array. We bespreken in hetvolgende hoofstuk waarom we een kopie van de `squares` array maken.

Reviewer 4: ```suggestion
Merk op hoe in `handleClick` we `.slice()` oproepen om een kopie van de `squares` array te modificeren in plaats van de bestaande array. We bespreken in het volgende hoofdstuk waarom we een kopie van de `squares` array maken.
```

### Waarom Immutability Belangrijk Is {#why-immutability-is-important}

> In het vorige code voorbeeld, hebben we aangeraden dat je de `.slice()` operator gebruikt om een kopie van de `squares` array creeert, om deze aan te passen. We bespreken immutability en waarom het belangrijk is om immutability te leren.

Reviewer 4: ```suggestion
In het vorige code voorbeeld, hebben we aangeraden dat je de `.slice()` operator gebruikt om een kopie van de `squares` array creëert, om deze aan te passen. We bespreken immutability en waarom het belangrijk is om immutability te leren.
```

// var newPlayer = {...player, score: 2};
```

Het eindresultaat is hetzelfde maar door het niet-muteren (of veranderen van onderliggende data) verkrijgen we enkele voordelen.

> #### Complex Features worden Simpel {#complex-features-become-simple}

Reviewer 4: ```suggestion
#### Complexe Features Worden Simpel {#complex-features-become-simple}
```

Immutability maakt complex features veel gemakkelijker om toe te passen. Later in deze tutorial zullen we een "time travel" feature toepassen dat ons toelaat de spel geschiedenis te bekijken en "terug te springen" naar een vorige zet. Deze functionaliteit is niet specifiek voor spelletjes -- de mogelijkheid om zekere acties ongedaan of terug te zetten is een normale vereiste in applicaties. Directe data mutatie vermijden laat ons toe verschillende versies van de spel geschiedenis intact te houden en later te gebruiken.

> #### Veranderingen detecteren {#detecting-changes}

Reviewer 4: ```suggestion
#### Veranderingen Detecteren {#detecting-changes}
```

> Veranderingen detecteren in mutable objects is moeilijk want ze worden direct gewijzigdt. Deze detectie vereist dat het mutable object zelf vergelijkt wordt met vorige kopiën van zichzelf en de volledige object tree die gevolgd werd.

Reviewer 4: ```suggestion
Veranderingen detecteren in mutable objects is moeilijk want ze worden direct gewijzigd. Deze detectie vereist dat het mutable object zelf vergelijkt wordt met vorige kopieën van zichzelf en de volledige object tree die gevolgd werd.
```

> Veranderingen detecteren in immutable objects is aanzienlijk gemakkelijker. Als het immutable object dat gerefereerd word anders is dan het vorige dan is het object veranderd.

Reviewer 4: ```suggestion
Veranderingen detecteren in immutable objects is aanzienlijk gemakkelijker. Als het immutable object dat gerefereerd wordt anders is dan het vorige dan is het object veranderd.
```

> #### Bepalen wanneer te herrenderen in React {#determining-when-to-re-render-in-react}

Reviewer 4: ```suggestion
#### Bepalen Wanneer Nogmaals Te Renderen In React {#determining-when-to-re-render-in-react}
```

> Het belangrijkste voordeel van immutability is dat het je helpt _pure components_ te maken in React. Immutable data kan gemakkelijk bepalen of veranderingen zijn gemaakt, wat op zich helpt determineren of een component her-rendert hoeft te worden.

Reviewer 4: ```suggestion
Het belangrijkste voordeel van immutability is dat het je helpt _pure components_ te maken in React. Immutable data kan gemakkelijk bepalen of veranderingen zijn gemaakt, wat op zich helpt te determineren of een component opnieuw gerenderd hoeft te worden.
```

Je kan meer leren over `shouldComponentUpdate()` en hoe *pure components* te bouwen door [Optimizing Performance](/docs/optimizing-performance.html#examples) te lezen.

> ### Function Components {#function-components}

Reviewer 4: ```suggestion
### Functiecomponenten {#function-components}
```

> We zullen nu Square verandere in een **function component**.

Reviewer 4: ```suggestion
We zullen nu Square veranderen in een **functiecomponent**.
```

> In React, zijn **function components** een simpelere manier om componenten aan te maken die enkel een `render` methode hebben en geen eigen state hebben. In plaats van een classe te definiere die `React.Component` extend, schrijven we een functie die `props` als input neemt en wat gerendered moet returned. Function components zijn minder langdradig om te schrijven dan classes en vele componenten kunnen zo geschreven worden.

Reviewer 4: ```suggestion
In React, zijn **functiecomponenten** een simpelere manier om componenten aan te maken die enkel een `render` methode hebben en geen eigen state hebben. In plaats van een class te definiëren die `React.Component` uitbreidt, schrijven we een functie die `props` als input neemt en wat gerenderd moet worden, teruggeeft. Functionele componenten zijn minder langdradig om te schrijven dan classes en veel componenten kunnen zo geschreven worden.
```

> Vervang Square class met deze functie:

Reviewer 4: ```suggestion
Vervang de Square class door deze functie:
```

}
```

We veranderden `this.props` in `props` beide keren dat het voorkwam.

**[De volledige code tot zover](https://codepen.io/gaearon/pen/QvvJOv?editors=0010)**

> >Nota

Reviewer 4: ```suggestion
>Opmerking
```

>

> >Wanneer we het Square component in een function component veranderden, veranderde we ook `onClick={() => this.props.onClick()}` in het kortere `onClick={props.onClick}` (merk op: het weglaten van haakjes aan *beide* kanten). In een klasse, gebruiken we een arrow function om toegang te krijgen tot de correcte `this` waarde, maar in een function component moeten we geen zorgen maken over `this`.

Reviewer 4: ```suggestion
>Wanneer we het Square component in een functiecomponent veranderden, veranderde we ook `onClick={() => this.props.onClick()}` in het kortere `onClick={props.onClick}` (merk op: het weglaten van haakjes aan *beide* kanten). In een klasse, gebruiken we een arrow function om toegang te krijgen tot de correcte `this` waarde, maar in een functioneel component moeten we ons geen zorgen maken over `this`.
```

> >Wanneer we het Square component in een function component veranderden, veranderde we ook `onClick={() => this.props.onClick()}` in het kortere `onClick={props.onClick}` (merk op: het weglaten van haakjes aan *beide* kanten). In een klasse, gebruiken we een arrow function om toegang te krijgen tot de correcte `this` waarde, maar in een function component moeten we geen zorgen maken over `this`.

Reviewer 4: ```suggestion
>Wanneer we het Square component in een functiecomponent veranderden, veranderde we ook `onClick={() => this.props.onClick()}` in het kortere `onClick={props.onClick}` (merk op: het weglaten van haakjes aan *beide* kanten). In een klasse, gebruiken we een arrow function om toegang te krijgen tot de correcte `this` waarde, maar in een functiecomponent moeten we ons geen zorgen maken over `this`.
```

> ### Beurten nemen {#taking-turns}

Reviewer 4: ```suggestion
### Beurten Nemen {#taking-turns}
```

  }
```

> Met deze verandering, kunnen "X" and "O" beurten nemen. Laten we ook de "status" text in de `render` van Board zodat het zal weergeven welke speler de volgende zet heeft:

Reviewer 4: ```suggestion
Met deze verandering, kunnen "X" and "O" beurten nemen. Laten we ook de "status" tekst in de `render` van Board zodat het zal weergeven welke speler de volgende zet heeft:
```

      // the rest has not changed
```

> Na deze veranderingen heb je normaal, dit Board component:

Reviewer 4: ```suggestion
Na deze veranderingen, zou je dit Board component moeten hebben:
```

}
```

**[De volledige code tot zover](https://codepen.io/gaearon/pen/KmmrBy?editors=0010)**

> ### Een winnaar aanduiden {#declaring-a-winner}

Reviewer 4: ```suggestion
### Een Winnaar Aanduiden {#declaring-a-winner}
```

}
```

> We roepen `calculateWinner(squares)` op in de `render` functie van Board om te controleren of een speler gewonnen heeft. Als een speler gewonnen heeft kunnen we text "Winner: X" of "Winner: O" weergeven. We verplacen de `status` declaratie in de `render` functie van Board met devolgende code:

Reviewer 4: ```suggestion
We roepen `calculateWinner(squares)` op in de `render` functie van Board om te controleren of een speler gewonnen heeft. Als een speler gewonnen heeft kunnen we de tekst "Winner: X" of "Winner: O" weergeven. We vervangen de `status` declaratie in de `render` functie van Board met de volgende code:
```

  }
```

**[De volledige code tot zover](https://codepen.io/gaearon/pen/LyyXgK?editors=0010)**

Gefeliciteerd! Je hebt nu een werkend tic-tac-toe spel. En je ook zojuist de basis van React geleerd. Dus eigenlijk ben *jij* de echte winnaar.

> ## Time Travel toevoegen {#adding-time-travel}

Reviewer 4: ```suggestion
## Time Travel Toevoegen {#adding-time-travel}
```

Als een laatste opgave, gaan we het mogelijk maken om "terug in de tijd" te gaan naar vorige zetten in het spel.

> ### Opslaan van de Geschiedenis van de Zetten in het Spel {#storing-a-history-of-moves}

Reviewer 4: ```suggestion
### De Geschiedenis Van Zetten Bijhouden {#storing-a-history-of-moves}
```

Als we de `squares` array muteren, zou het implementeren van time travel zeer moeilijk zijn.

Maar in plaats hiervan hebben we `slice()` gebruikt om een nieuwe kopie van de `squares` array aan te maken na elke zet en aanschouwen we het als [immutable](#why-immutability-is-important). Dit laat ons toe om elke vorige versie van het `squares` array op te slaan en te navigeren naar alle verleden zetten.

> We zullen de vorige `squares` arrays opslaan in een andere array `history`. De `history` array representeerd alle states van het Board component. Van de eerste tot laatste zet, en ziet er zo uit:

Reviewer 4: ```suggestion
We zullen de vorige `squares` arrays opslaan in een andere array `history`. De `history` array representeert alle states van het Board component. Van de eerste tot laatste zet, en ziet er zo uit:
```

]
```

Nu moeten we beslissen welk component de `history` state bevat.

> ### De staat verheffen (lifting state), nogmaals {#lifting-state-up-again}

Reviewer 4: ```suggestion
### Nogmaals De State Optillen {#lifting-state-up-again}
```

We willen dat het top-level Game component een lijst van vorige zetten toont.
Hiervoor heeft het component toegang nodig tot `history` en dus zetten we de `history` state in het top-level Game component.

Het plaatsen van de `history` state in het Game component laat ons toe om de `squares` state te verwijderen van het child component, Board. Net [zoals we de state verheften ("lifted state up")](#lifting-state-up) van het Square component naar het Board component, gaan we het nu verheffen van het Board element naar het top-level Game component. Dit geeft het Game component volledige controle over de data van Board en laat het toe om het Board component vorige zetten te renderen.

> Eerst, we zetten de initiële state van het Game component in zijn constructor:

Reviewer 4: ```suggestion
Eerst zetten we de initiële state van het Game component in zijn constructor:
```

  }
```

> >Nota

Reviewer 4: ```suggestion
>Opmerking
```

>

> > Tot tegenstelling van de array `push()` methode, waar je misschien meer vertrouwd met bent, muteerd de `concat()` methode de originele array niet, en wordt daarom verkozen.

Reviewer 4: ```suggestion
>In tegenstelling tot de array `push()` methode, waar je misschien meer vertrouwd mee bent, muteert de `concat()` methode de originele array niet, en wordt daarom verkozen.
```

Op dit moment heeft het Board component enkel de `renderSquare` en `render` methodes nodig. De staat van het spel en de `handleClick` methode zouden in het Game component zijn.

**[De volledige code tot zover](https://codepen.io/gaearon/pen/EmmOqJ?editors=0010)**

> ### Tonen van Vorige Zetten {#showing-the-past-moves}

Reviewer 4: ```suggestion
### Vorige Zetten Tonen {#showing-the-past-moves}
```

Aangezien we de geschiedenis van het tic-tac-toe spel opnemen, kunnen we nu een lijst van vorige zetten aan de speler tonen.

> We hebben eerder geleerd dat React elementen eerste klasse Javascript objecten zijn; we kunenn ze doorgeven in onze applicaties. Om meerdere items te renderen kunnen we een array van elementen gebruiken.

Reviewer 4: ```suggestion
We hebben eerder geleerd dat React elementen eerste klasse JavaScript objecten zijn; we kunnen ze doorgeven in onze applicaties. Om meerdere items te renderen kunnen we een array van elementen gebruiken.
```

> In Javascript hebben arrays een [`map()` methode](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map) dat doorgaans gebruikt wordt om data naar andere data te mappen, bijvoorbeeld:

Reviewer 4: ```suggestion
In JavaScript hebben arrays een [`map()` methode](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map) dat doorgaans gebruikt wordt om data naar andere data te mappen, bijvoorbeeld:
```

  }
```

**[De volledige code tot zover](https://codepen.io/gaearon/pen/EmmGEa?editors=0010)**

Voor elke zet in de tic-tac-toe spel geschiedenis, creëeren we een list item `<li>` die een knop `<button>` bevat. De knop heeft een `onClick` handler die de `this.jumpTo()` methode oproept. We hebben de `this.jumpTo()` methode nog niet geïmplementeerd. Tot zover zien we een lijst van zetten die zijn gemaakt in het spel en een waarschuwing in de developer tools console:

>  Warning:
>  Each child in an array or iterator should have a unique "key" prop. Check the render method of "Game".

> Laten we bespreken wat deze waarschuwing betekend.

Reviewer 4: ```suggestion
Laten we bespreken wat deze waarschuwing betekent.
```

> ### Picking a Key {#picking-a-key}

Reviewer 4: ```suggestion
### Een Key Kiezen {#picking-a-key}
```

> Wanner we een lijst renderen, slaat React informatie op over elk gerendered element (list element). Wanneer we een llijst updaten, moet React beslissen wat veranderd moet worden. We zouden lijst elementen kunnen ingevoeg, verwijderen, herordenen of updaten.

Reviewer 4: ```suggestion
Wanneer we een lijst renderen, slaat React informatie op over elk gerenderd element (list element). Wanneer we een lijst updaten, moet React beslissen wat veranderd moet worden. We zouden lijst elementen kunnen invoegen, verwijderen, herordenen of updaten.
```

> Beeld je in om overtegaan van:

Reviewer 4: ```suggestion
Beeld je in om over te gaan van:
```

```html
<li>Alexa: 7 tasks left</li>
<li>Ben: 5 tasks left</li>
```

naar

```html
<li>Ben: 9 tasks left</li>
<li>Claudia: 8 tasks left</li>
<li>Alexa: 5 tasks left</li>
```

> Naast de geupdate getallen, zou een mens die dit leest waarschijnlijk beweren dat we de rangschikking van Alexa en Ben veranderd hebben, en Claudia tussen Alexa en Ben hebben ingevoegd.

Reviewer 4: ```suggestion
Naast de geüpdatete getallen, zou een mens die dit leest waarschijnlijk beweren dat we de rangschikking van Alexa en Ben veranderd hebben, en Claudia tussen Alexa en Ben hebben ingevoegd.
```

`key` is een speciaal en gereserveerd property in React (samen met `ref`, een geavanceerde feature). Wanneer een element gecreëerd wordt, leidt React de `key` property af en slaat de key dadelijk op in het gereturnde element. Hoewel het lijkt alsof `key` behoort in de `props`, kunnen we `this.props.key` niet gebruiken om naar `key` te refereren. React gebruikt `key` automatisch om te beslissen welke componenten up te daten. Een component kan niet vragen naar zijn `key`.

**Het is sterk aangeraden dat je gepaste keys toeschrijft wanneer je dynamische lijsten aanmaakt.** Als je geen gepaste keys hebt, kan je overwegen om je data te herschikken totdat je er wel hebt.

Als geen key wordt bepaald, zal React een waarschuwing tonen en bij verstek de array index gebruiken. De array index als key gebruiken is problematisch wanneer je de lijst elementen probeert te herordenen of elementen invoegt/verwijderd. Expliciet `key={i}` toewijzen verzwijgt de waarschuwing maar heeft dezelfde problemen als de array indexen en is niet aan te raden in de meeste toepassingen.

> keys hoeven niet globaal uniek te zijn; ze moeten enkel uniek zijn ten opzichte van componenten en hun siblings.

Reviewer 4: ```suggestion
Keys hoeven niet globaal uniek te zijn; ze moeten enkel uniek zijn ten opzichte van componenten en hun siblings.
```

  }
```

> We maken nu enkele veranderingen aan de Game `handleClick` methode, opgeroept wordt wanneer je een Square aanklikt.

Reviewer 5: opgeroept => voltooid deelwoord van oproepen in opgeroepen
maar aanroepen is hier beter
```suggestion
We maken nu enkele veranderingen aan de `handleClick` methode van Game, die aangeroepen wordt wanneer je een Square aanklikt.
```
